### PR TITLE
Enable Lightning tests and sort out logging inside e2e tests

### LINKED ISSUE #2192: Fix LND-based sanity tests
_Follow-up ticket of https://github.com/comit-network/comit-rs/pull/2106_

## Problem

This PR introduces skips LND sanity issues. This should not be the case.
It currently timeout at `genSeed` phase, sometimes it works.

- lnd is slow to catch-up with blocks from wallet initialisation
- 2 steps to init wallet: `genSeed` (to generate seed words) and then `initWallet` using the previous seed.
- `genSeed` accepts a byte array as entropy, this can be pseudo random to not use as much entropy on the local machine.

## Goal
Have a stable and preferable fast (parallelizable) LN e2e test setup. 

## Recommendation
- Start LND nodes in the test setup not within the tests so that they can be shared between tests.

### LINKED ISSUE #2221: No more `console.log` in the e2e tests
## Problem

We still have `console.log` statements around in the e2e tests from prior to moving to jest. We now no longer control the commandline arguments so we have this temporary hack for detecting the verbose flag: https://docs.rs/futures/0.3.4/futures/future/trait.FutureExt.html#method.now_or_never

Additionally, we are no longer outputting any information on stdout but all the info goes into log files. This should also be the case for the current `console.log` statements.

## Goal

No hacks in our code and proper logging.

## Recommendation

We can now easily pass around a logger instance.
We can just use the logger instance for logging all the information to files.

As part of this issue, we probably want to generally clean up the logging in the e2e tests. There is some mess with how/when we pass specific loggers.

The multi-file appender might of use for this: https://github.com/log4js-node/log4js-node/blob/master/docs/multiFile.md

### PR DESCRIPTION
**Pull request by thomaseizinger**
First, I am sorry for scope creeping this PR a bit but if you hear me out, I can explain what happened :D

So first, I set out to make the lnd tests work. Judging from the first couple of patches, you can see that this was achieved fairly quickly. The major breakthrough was 559bb4f6b0900735a865a328b5952758427e6ae3 which passes 16 predefined bytes as the seed to LND.

As you can see from the patch 3ebc309d8aea6ca9b3a5b6b6609e163d05ca6dcc, moving the startup of LND to the test environment was quite a pain due to the logger not being available from within the test environment.

Also to make it actually work, I had to start the ledgers sequentially: b5d0e8812c2116239eb1ba2ee2ad6d1b34fc8065

From then on, the tests passed consistently and I started to clean up the mess I caused. The remaining patches are basically that :)

I've made an effort in nicely slicing the patches and would recommend to review the PR like this to get an idea, which changes are related to each other.
Overall, I believe these patches are a massive improvement to our e2e tests.

All the ledgers are now started with as much parallelism as possible from within the test environment. Logging is now consistently in files inside the `log` directory. No more console outputs.

Resolves #2192.
Resolves #2221.

### CODE REVIEW

File: api_tests/lib/ledgers/ethereum.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * blockchain is up and running.
+ *
+ * This class serves as an abstraction layer on top of Ethereum, regardless
+ * of which implementation is used (Docker container, parity, geth, etc).
+ */
+export default class EthereumLedger implements LedgerInstance {
+    public static async start(instance: EthereumInstance) {
+        await instance.start();
+
+        await instance.start();
```
Comment (tcharding) [inline api_tests/lib/ledgers/ethereum.ts]: Duplicate line.

Comment (thomaseizinger) [reply api_tests/lib/ledgers/ethereum.ts]: Ah yeah that was an oversight that I fixed somewhere later :)

Gonna patch it out and force push!

**Comment by thomaseizinger** [comment]
@D4nte I would like to have your review on this as-well because you wrote the first LND integration in the e2e tests and also had some thoughts during the grooming!

**Comment by D4nte** [comment]
> @D4nte I would like to have your review on this as-well because you wrote the first LND integration in the e2e tests and also had some thoughts during the grooming!

I am working on it, will be done by 12:00

**Comment by thomaseizinger** [comment]
> Really good stuff. The review was much easier with the separate commits! Would you like to squash them before merging though? I applied the `no-mergify` label just in case.

I was thinking about leaving them, they are atomic in the sense that each one works :) (hopefully :crossed_fingers: )

**Comment by thomaseizinger** [comment]
> > @D4nte I would like to have your review on this as-well because you wrote the first LND integration in the e2e tests and also had some thoughts during the grooming!
> 
> I am working on it, will be done by 12:00

Cheers, feel free to just remove the `no-mergify` then if you don't have any objections!

File: api_tests/lib/wallets/ethereum.ts [hunk 2]
```diff
@@ -107,17 +106,9 @@ export class EthereumWallet implements Wallet {
         return this.inner.getAccount();
     }
 
-    public async deployErc20TokenContract(
-        projectRoot: string
-    ): Promise<string> {
+    public async deployErc20TokenContract(): Promise<string> {
         const data =
```
Comment (D4nte) [inline api_tests/lib/wallets/ethereum.ts]: Could you actually put it at the end of the file in a const to help readability?

File: api_tests/lib/create_actors.ts [hunk 3]
```diff
@@ -13,15 +12,12 @@ export async function createActors(
 
     const listPromises: Promise<Actor>[] = [];
     for (const name of actorNames) {
-        const cndLogFile = path.join(
-            global.logRoot,
+        const cndLogFile = global.getLogFile([
```
Comment (D4nte) [inline api_tests/lib/create_actors.ts]:
```suggestion
        const cndLogFile = global.getLogDir([
```
This returns the path to the `logs` dir, not file.

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 4]
```diff
@@ -132,28 +154,35 @@ export class LndInstance {
...
 
-    public getGrpcPort() {
-        return this.actorConfig.lndRpcPort;
+    get grpcPort() {
+        return this.lndRpcPort;
     }
 
-    public getLightningSocket() {
-        return `${this.getLightningHost()}:${this.getLightningPort()}`;
+    get p2pSocket() {
```
Comment (D4nte) [inline api_tests/lib/ledgers/lnd_instance.ts]:
```suggestion
    get lightningSocket() {
```
Best to keep same terminology everywhere (see `lightningHost` and `lightningPort` below)

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 5]
```diff
@@ -132,28 +154,35 @@ export class LndInstance {
...
 
-    public getLightningPort() {
-        return this.actorConfig.lndP2pPort;
+    get lightningPort() {
+        return this.lndP2pPort;
+    }
+
+    get config(): LightningNodeConfig {
+        return {
+            p2pSocket: this.p2pSocket,
```
Comment (D4nte) [inline api_tests/lib/ledgers/lnd_instance.ts]:
```suggestion
            lightningSocket: this.lightningSocket,
```

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 6]
```diff
@@ -1,34 +1,47 @@
...
+
+    private constructor(
+        private readonly dataDir: string,
+        private readonly name: string,
         private readonly logger: Logger,
-        private readonly testLogDir: string,
-        private readonly actorConfig: E2ETestActorConfig,
-        private readonly bitcoindDataDir: string
+        private readonly bitcoindDataDir: string,
+        private readonly lndP2pPort: number,
```
Comment (D4nte) [inline api_tests/lib/ledgers/lnd_instance.ts]: Just choose between `p2p` and `lightning` please.

File: api_tests/lib/ledgers/parity_instance.ts [hunk 7]
```diff
@@ -1,27 +1,48 @@
...
+        projectRoot: string,
+        logFile: string,
+        logger: Logger
+    ) {
+        return new ParityInstance(
+            projectRoot,
+            logFile,
+            logger,
+            await getPort({ port: 8545 }),
+            await getPort()
```
Comment (D4nte) [inline api_tests/lib/ledgers/parity_instance.ts]:
```suggestion
            await getPort({ port: 30303 })
```

File: api_tests/lib/wallets/bitcoin.ts [hunk 8]
```diff
@@ -48,10 +50,14 @@ export class BitcoinWallet implements Wallet {
             );
         }
 
+        const amount = toBitcoin(minimumExpectedBalance.times(2).toString());
+
         await this.bitcoinRpcClient.sendToAddress(
             toAddress,
-            toBitcoin(minimumExpectedBalance.times(2).toString()) // make sure we have at least twice as much
+            amount // make sure we have at least twice as much
```
Comment (D4nte) [inline api_tests/lib/wallets/bitcoin.ts]: Move the comment where `.times(2)` is done.

File: api_tests/lib/wallets/lightning.ts [hunk 9]
```diff
@@ -14,7 +13,7 @@ import { AddressType } from "@radar/lnrpc";
 export class LightningWallet implements Wallet {
     public static async newInstance(
         bitcoinWallet: BitcoinWallet,
-        logger: Logger,
+        // logger: Logger,
```
Comment (D4nte) [inline api_tests/lib/wallets/lightning.ts]:
```suggestion
```

File: api_tests/lib/wallets/lightning.ts [hunk 10]
```diff
@@ -25,16 +24,16 @@ export class LightningWallet implements Wallet {
             lndp2pSocket
         );
 
-        logger.debug("lnd getinfo:", await inner.lnd.lnrpc.getInfo());
+        // logger.debug("lnd getinfo:", await inner.lnd.lnrpc.getInfo());
```
Comment (D4nte) [inline api_tests/lib/wallets/lightning.ts]:
```suggestion
```

File: api_tests/lib/wallets/lightning.ts [hunk 11]
```diff
@@ -25,16 +24,16 @@ export class LightningWallet implements Wallet {
...
-        return new LightningWallet(inner, logger, bitcoinWallet);
+        return new LightningWallet(inner, bitcoinWallet);
     }
 
     public MaximumFee = 0;
 
     private constructor(
         public readonly inner: LightningWalletSdk,
-        private readonly logger: Logger,
+        // private readonly logger: Logger,
```
Comment (D4nte) [inline api_tests/lib/wallets/lightning.ts]:
```suggestion
```

File: api_tests/lib/wallets/lightning.ts [hunk 12]
```diff
@@ -48,10 +47,10 @@ export class LightningWallet implements Wallet {
         const startingBalance = new BigNumber(
             await this.getBalanceByAsset(asset)
         );
-        this.logger.debug("starting: ", startingBalance.toString());
+        // this.logger.debug("starting: ", startingBalance.toString());
```
Comment (D4nte) [inline api_tests/lib/wallets/lightning.ts]:
```suggestion
```

File: api_tests/lib/wallets/lightning.ts [hunk 13]
```diff
@@ -48,10 +47,10 @@ export class LightningWallet implements Wallet {
...
 
         const minimumExpectedBalance = new BigNumber(asset.quantity);
-        this.logger.debug("min expected: ", minimumExpectedBalance.toString());
+        // this.logger.debug("min expected: ", minimumExpectedBalance.toString());
```
Comment (D4nte) [inline api_tests/lib/wallets/lightning.ts]:
```suggestion
```

File: api_tests/lib/ledgers/parity_instance.ts (hunk 7 above)
Comment (thomaseizinger) [reply api_tests/lib/ledgers/parity_instance.ts]: I don't see the point in doing this. I also don't know why we are doing this everywhere else. I just want _some_ port, don't care which one it is.

File: api_tests/lib/wallets/lightning.ts (hunk 9 above)
Comment (thomaseizinger) [reply api_tests/lib/wallets/lightning.ts]: Ups :D

Yeah those should be activated again :)

**Comment by thomaseizinger** [comment]
@D4nte 

> However I cannot approve as it fails locally on my machine.

CI is fine and my machine too :man_shrugging: 
The error you are getting sounds node-version related thought?

**Comment by D4nte** [comment]
> @D4nte
> 
> > However I cannot approve as it fails locally on my machine.
> 
> CI is fine and my machine too 🤷‍♂
> The error you are getting sounds node-version related thought?

Should the node engine version in `package.json` be more restrictive? What node version do you use?

File: api_tests/lib/ledgers/lnd_instance.ts (hunk 6 above)
Comment (thomaseizinger) [reply api_tests/lib/ledgers/lnd_instance.ts]: Going for p2p for consistency with the bitcoin code.

File: api_tests/lib/wallets/ethereum.ts (hunk 2 above)
Comment (thomaseizinger) [reply api_tests/lib/wallets/ethereum.ts]: Do you have word wrap turned on or what is the issue?

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 14]
```diff
@@ -4,6 +4,7 @@ import * as path from "path";
 import getPort from "get-port";
 import { LogReader } from "./log_reader";
 import { Lnd } from "comit-sdk";
+import whereis from "@wcjiang/whereis";
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply api_tests/lib/ledgers/lnd_instance.ts]: We assumed `lnd` is in the path, whereas everywhere else we pass a path to it.

The error message was ugly when we just started `lnd`, so I am using this to find it in the path which errors out much earlier.

File: api_tests/lib/create_actors.ts (hunk 3 above)
Comment (thomaseizinger) [reply api_tests/lib/create_actors.ts]: Negative, this just returns a path and is supposed to be used for files and not directories.

**Comment by D4nte** [comment]
Are the thousands `RUNS  ...` messages appearing in the console when running the e2e tests new with this PR?

File: api_tests/lib/ledgers/parity_instance.ts (hunk 7 above)
Comment (D4nte) [reply api_tests/lib/ledgers/parity_instance.ts]: To me this port preference looked new and introduced by this PR hence my proposal. I did not realize it was not introduce by this PR.

**Comment by thomaseizinger** [comment]
> Are the thousands `RUNS ...` messages appearing in the console when running the e2e tests new with this PR?

I am not sure I know what you are talking about?

File: api_tests/lib/wallets/ethereum.ts (hunk 2 above)
Comment (D4nte) [reply api_tests/lib/wallets/ethereum.ts]: I just think that a giant variable is better at the end of a test file.

**Comment by D4nte** [comment]
> > Are the thousands `RUNS ...` messages appearing in the console when running the e2e tests new with this PR?
> 
> I am not sure I know what you are talking about?

Ok it's a CLion odd thing. All good.

**Pull request merged**